Aggressive pruner AssertionError

Running a function like:
```
def annotate_tx_expression_data(ht, tx_ht, location):
    key = ht.key if isinstance(ht, hl.Table) else ht.row_key
    return hl.find(lambda csq: (csq.ensg == location.gene_id) &
                               (csq.csq == location.most_severe_consequence),
                   tx_ht[key].tx_annotation)
mt = mt.annotate_rows(expressed=annotate_tx_expression_data(mt, tx_ht, mt.lof_csqs).mean_expression > 0.1)
mt.describe()
mt.group_rows_by(*list(annotation_expr.keys())).aggregate_rows(
        classic_caf=hl.agg.sum(mt.freq[0].AF),
        max_af=hl.agg.max(mt.freq[0].AF),
        classic_caf_array=hl.agg.array_sum(mt.freq.map(lambda x: x.AF))
    ).aggregate_entries(
        num_homs=hl.agg.count_where(mt.GT.is_hom_var()),
        num_hets=hl.agg.count_where(mt.GT.is_het()),
        defined_sites=hl.agg.count_where(hl.is_defined(mt.GT))
    ).result()
```
errors out with `Error summary: AssertionError: assertion failed: ensg not in struct{mean_expression: float64}`. the describe shows that it's doing the right thing (`expressed` is a `bool`), but i'm guessing that since ensg is not strictly referred to except in a lambda, that it's getting pruned out?

Full log posted on zulip

--- a/hail/src/main/scala/is/hail/expr/ir/IR.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/IR.scala
@@ -79,11 +79,11 @@ final case class NA(typ: Type) extends IR { assert(!typ.required) }
 final case class IsNA(value: IR) extends IR { val typ = TBoolean() }
 
 object If {
-  def unify(cond: IR, cnsq: IR, altr: IR): If = {
+  def unify(cond: IR, cnsq: IR, altr: IR, unifyType: Option[Type] = None): If = {
     if (cnsq.typ == altr.typ)
       If(cond, cnsq, altr)
     else {
-      val t = cnsq.typ.deepOptional()
+      val t = unifyType.getOrElse(cnsq.typ)
       If(cond,
         PruneDeadFields.upcast(cnsq, t),
         PruneDeadFields.upcast(altr, t))

--- a/hail/src/main/scala/is/hail/expr/ir/IR.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/IR.scala
@@ -83,7 +83,7 @@ object If {
     if (cnsq.typ == altr.typ)
       If(cond, cnsq, altr)
     else {
-      val t = unifyType.getOrElse(cnsq.typ)
+      val t = unifyType.getOrElse(cnsq.typ.deepOptional())
       If(cond,
         PruneDeadFields.upcast(cnsq, t),
         PruneDeadFields.upcast(altr, t))

--- a/hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala
@@ -952,7 +952,7 @@ object PruneDeadFields {
         val cond2 = rebuild(cond, in, memo)
         val cnsq2 = rebuild(cnsq, in, memo)
         val alt2 = rebuild(alt, in, memo)
-        If.unify(cond2, cnsq2, alt2)
+        If.unify(cond2, cnsq2, alt2, unifyType = Some(requestedType))
       case Let(name, value, body) =>
         val value2 = rebuild(value, in, memo)
         Let(

--- a/hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala
@@ -48,7 +48,7 @@ object TypeCheck {
         check(cnsq)
         check(altr)
         assert(cond.typ.isOfType(TBoolean()))
-        assert(cnsq.typ == altr.typ, s"${ cnsq.typ }, ${ altr.typ }, $cond")
+        assert(cnsq.typ == altr.typ, s"Type mismatch:\n  cnsq: ${ cnsq.typ.parsableString() }\n  altr: ${ altr.typ.parsableString() }\n  $x")
         assert(x.typ == cnsq.typ)
 
       case x@Let(name, value, body) =>

--- a/hail/src/test/scala/is/hail/expr/ir/PruneSuite.scala
+++ b/hail/src/test/scala/is/hail/expr/ir/PruneSuite.scala
@@ -876,5 +876,22 @@ class PruneSuite extends SparkSuite {
         ir.child.typ == subsetMatrixTable(mr.typ, "va.r2")
       })
   }
+
+  @Test def testIfUnification() {
+    val pred = False()
+    val t = TStruct("a" -> TInt32(), "b" -> TInt32())
+    val pruneT = TStruct("a" -> TInt32())
+    val cnsq = Ref("x", t)
+    val altr = NA(t)
+    val ifIR = If(pred, cnsq, altr)
+    val memo = Memo.empty[BaseType]
+      .bind(pred, TBoolean())
+      .bind(cnsq, pruneT)
+      .bind(altr, pruneT)
+      .bind(ifIR, pruneT)
+
+    // should run without error!
+    PruneDeadFields.rebuild(ifIR, Env.empty[Type].bind("a", t), memo)
+  }
 }
 

